File: packages/application/src/delete-client.use-case.ts
import {
  ClientHasInvoicesError,
  ClientNotFoundError,
  type IClientRepository,
  type IInvoiceRepository,
  type ILogger,
} from '@monolegal/domain';

export class DeleteClientUseCase {
  constructor(
    private readonly clientRepository: IClientRepository,
    private readonly invoiceRepository: IInvoiceRepository,
    private readonly logger: ILogger,
  ) {}

  async execute(id: string): Promise<void> {
    const existing = await this.clientRepository.findById(id);
    if (!existing) {
      throw new ClientNotFoundError(`Client not found: ${id}`);
    }

    const invoices = await this.invoiceRepository.findByClientId(id);
    if (invoices.length > 0) {
      throw new ClientHasInvoicesError(
        `Cannot delete client ${id}: ${invoices.length} invoice(s) exist`,
      );
    }

    await this.clientRepository.delete(id);
    this.logger.info('Client deleted', { clientId: id });
  }
}

File: backend/dotnet/src/Monolegal.Application/DeleteClientUseCase.cs
using Monolegal.Domain.Errors;
using Monolegal.Domain.Ports;

namespace Monolegal.Application;

public class DeleteClientUseCase(
    IClientRepository clientRepository,
    IInvoiceRepository invoiceRepository,
    ILogger logger)
{
    public async Task ExecuteAsync(string id, CancellationToken cancellationToken = default)
    {
        var existing = await clientRepository.FindByIdAsync(id, cancellationToken);
        if (existing is null)
            throw new ClientNotFoundError($"Client not found: {id}");

        var invoices = await invoiceRepository.FindByClientIdAsync(id, cancellationToken);
        if (invoices.Count > 0)
            throw new ClientHasInvoicesError($"Cannot delete client {id}: {invoices.Count} invoice(s) exist");

        await clientRepository.DeleteAsync(id, cancellationToken);
        logger.Info("Client deleted", new Dictionary<string, object?> { ["clientId"] = id });
    }
}
